### PR TITLE
ARROW-1237: [JAVA] expose the ability to set lastSet

### PR DESCRIPTION
**Pull request by siddharthteotia**
The changes here expose the ability to set "lastSet" on Nullable<var length>Vector. 
I believe this is needed only for NullableVarCharVector and NullableVarBinaryVector. Hence the API is exposed through NullableValueVectors.java

### CODE REVIEW

**Comment by wesm** [comment]
Is it possible to add a unit test for this? Can you also indicate in the PR title that this is a Java patch, e.g. `[Java]` or `[JAVA]`?

**Comment by siddharthteotia** [comment]
Edited the PR title to include [JAVA].
Let me get back after giving a thought on unit-tests

File: java/vector/src/main/codegen/templates/NullableValueVectors.java [hunk 1]
```diff
@@ -699,6 +699,14 @@ public void reset(){
       setCount = 0;
       <#if type.major = "VarLen">lastSet = -1;</#if>
     }
+
+    public void setLastSet(int value) {
+      <#if type.major = "VarLen">lastSet = value;</#if>
```
Comment (DremioQA) [inline java/vector/src/main/codegen/templates/NullableValueVectors.java]: Is there a lastset for the list vector as well?

**Review by DremioQA — CHANGES_REQUESTED** [review]
Let's add some unit tests for each version that has last set. Basically, set some data via memory access then update last set, then confirm a correct vector output.

File: java/vector/src/main/codegen/templates/NullableValueVectors.java (hunk 1 above)
Comment (siddharthteotia) [inline java/vector/src/main/codegen/templates/NullableValueVectors.java]: Writing the unit tests. 

Yes, the ListVector also has lastSet. Will add the API for ListVector as well and unit tests for each vector type; NullableVarCharVector, NullableVarBinaryVector, ListVector.

Comment (siddharthteotia) [inline java/vector/src/main/codegen/templates/NullableValueVectors.java]: @DremioQA

**Comment by siddharthteotia** [comment]
@jacques-n 
I have added some unit tests to ensure correctness of vector data with the usage of setLastSet(). 

There is a slightly orthogonal test case test based on VectorLoader and VectorUnloader. 
If it looks redundant, I will remove it. It was written since my initial understanding on loading a vector was based on VectorLoader and VectorUnloader. Probably no harm to leave it since it still tests the loading-unloading of Nullable<var length> vector.

File: java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java [hunk 2]
```diff
@@ -509,11 +513,105 @@ public void testCopyFromWithNulls() {
...
+      mutator1.set(1, STR2);
+      mutator1.set(2, STR3);
+      mutator1.setValueCount(3);
+      assertEquals(Integer.toString(2), Integer.toString(mutator1.getLastSet()));
+
+      /* Check the vector output */
+      final NullableVarCharVector.Accessor accessor1 = vector1.getAccessor();
+      assertArrayEquals(STR1, accessor1.get(0));
+      assertArrayEquals(STR2, accessor1.get(1));
+      assertArrayEquals(STR3, accessor1.get(2));
```
Comment (jacques-n) [inline java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java]: Given the main issue of holes/blanks in the items set, I suggest you modify these tests to set the valuecount to something higher (e.g. 15). Then verify that everything looks good.

Comment (siddharthteotia) [inline java/vector/src/test/java/org/apache/arrow/vector/TestValueVector.java]: Added additional test and slightly modified existing ones.

**Pull request closed without merging**